File: app/src/test/java/com/rxjava/example/java/TransformingObservables.java
package com.rxjava.example.java;

import io.reactivex.functions.BiFunction;
import org.junit.Test;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;

public class TransformingObservables {

    @Test
    public void rxBuffer() {
        Observable.just(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
                .buffer(3)
                .subscribe(new Observer<List<Integer>>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        System.out.println("onSubscribe");
                    }

                    @Override
                    public void onNext(List<Integer> integers) {
                        System.out.println("onNext: " + integers);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        System.out.println("onComplete");
                    }
                });
    }

    @Test
    public void rxFlatMap() {
        Observable<Integer> observable1 = Observable.just(1, 2, 3);
        final Observable<Integer> observable2 = Observable.just(1, 2, 3);

        observable1.flatMap(new Function<Integer, ObservableSource<Integer>>() {
            @Override
            public ObservableSource<Integer> apply(Integer integer) throws Exception {
                return observable2;
            }
        }).subscribe(new Observer<Integer>() {
            @Override
            public void onSubscribe(Disposable d) {
                System.out.println("onSubscribe");
            }

            @Override
            public void onNext(Integer integer) {
                System.out.println("onNext: " + integer);
            }

            @Override
            public void onError(Throwable e) {

            }

            @Override
            public void onComplete() {
                System.out.println("onComplete");
            }
        });
    }

    @Test
    public void rxMap() {
        Observable.just(1, 2, 3, 4, 5)
                .map(new Function<Integer, Integer>() {
                    @Override
                    public Integer apply(Integer integer) throws Exception {
                        return integer * 2;
                    }
                })
                .subscribe(new Observer<Integer>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        System.out.println("onSubscribe");
                    }

                    @Override
                    public void onNext(Integer integer) {
                        System.out.println("onNext: " + integer);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        System.out.println("onComplete");
                    }
                });
    }

    @Test
    public void rxScan() {
        Observable.just(1, 2, 3, 4, 5)
                .scan(new BiFunction<Integer, Integer, Integer>() {
                    @Override
                    public Integer apply(Integer integer, Integer integer2) throws Exception {
                        return integer + integer2;
                    }
                })
                .subscribe(new Observer<Integer>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        System.out.println("onSubscribe");
                    }

                    @Override
                    public void onNext(Integer integer) {
                        System.out.println("onNext: " + integer);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        System.out.println("onComplete");
                    }
                });
    }

    @Test
    public void rxWindow1() {
        Observable.just(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
                .window(3)
                .subscribe(new Observer<Observable<Integer>>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        System.out.println("onSubscribe rxWindow1");
                    }

                    @Override
                    public void onNext(Observable<Integer> integerObservable) {
                        System.out.println("onNext rxWindow1: " + integerObservable);
                        rxWindow2(integerObservable);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        System.out.println("onComplete rxWindow1");
                    }
                });
    }

    private void rxWindow2(Observable<Integer> integerObservable) {
        integerObservable
                .map(new Function<Integer, Integer>() {
                    @Override
                    public Integer apply(Integer integer) throws Exception {
                        return integer;
                    }
                })
                .subscribe(new Observer<Integer>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        System.out.println("onSubscribe rxWindow2");
                    }

                    @Override
                    public void onNext(Integer integer) {
                        System.out.println("onNext rxWindow2: " + integer);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        System.out.println("onComplete rxWindow2");
                    }
                });
    }

}

File: app/src/test/java/com/rxjava/example/kotlin/TransformingObservables.kt
package com.rxjava.example.kotlin

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Function
import org.junit.Test

class TransformingObservables {

    @Test
    fun rxBuffer() {
        Observable.just(1, 2, 3, 4, 5, 6, 7)
            .buffer(3)
            .subscribe(object : Observer<List<Int>> {
                override fun onSubscribe(d: Disposable) {
                    println("onSubscribe")
                }

                override fun onNext(integers: List<Int>) {
                    println("onNext: $integers")
                }

                override fun onError(e: Throwable) {

                }

                override fun onComplete() {
                    println("onComplete")
                }
            })
    }

    @Test
    fun rxFlatMap() {
        val observable1 = Observable.just(1, 2, 3)
        val observable2 = Observable.just(1, 2, 3)

        observable1.flatMap(object : Function<Int, ObservableSource<Int>> {
            @Throws(Exception::class)
            override fun apply(integer: Int): ObservableSource<Int> {
                return observable2
            }
        }).subscribe(object : Observer<Int> {
            override fun onSubscribe(d: Disposable) {
                println("onSubscribe")
            }

            override fun onNext(integer: Int) {
                println("onNext: " + integer)
            }

            override fun onError(e: Throwable) {

            }

            override fun onComplete() {
                println("onComplete")
            }
        })
    }

    @Test
    fun rxMap() {
        Observable.just(1, 2, 3, 4, 5)
            .map(object : Function<Int, Int> {
                @Throws(Exception::class)
                override fun apply(integer: Int): Int {
                    return integer * 2
                }
            })
            .subscribe(object : Observer<Int> {
                override fun onSubscribe(d: Disposable) {
                    println("onSubscribe")
                }

                override fun onNext(integer: Int) {
                    println("onNext: " + integer)
                }

                override fun onError(e: Throwable) {

                }

                override fun onComplete() {
                    println("onComplete")
                }
            })
    }

    @Test
    fun rxMapLambda() {
        Observable.just(1, 2, 3, 4, 5)
            .map { integer -> integer * 2 }
            .subscribe(
                { next ->
                    println("onNext: $next")
                },
                { error ->
                    println(error.message.toString())
                },
                {
                    println("onComplete")
                }
            )
    }

    @Test
    fun rxScan() {
        Observable.just(1, 2, 3, 4, 5)
            .scan(object : BiFunction<Int, Int, Int> {
                @Throws(Exception::class)
                override fun apply(integer: Int, integer2: Int): Int {
                    return integer + integer2
                }
            })
            .subscribe(object : Observer<Int> {
                override fun onSubscribe(d: Disposable) {
                    println("onSubscribe")
                }

                override fun onNext(integer: Int) {
                    println("onNext: $integer")
                }

                override fun onError(e: Throwable) {

                }

                override fun onComplete() {
                    println("onComplete")
                }
            })
    }

}